only add the assemblers we actually need

jdannberg opened the pull request:
This builds on #5233. Instead of adding all assemblers to all fields, we now add only those we actually need. 
The tests will show if I have forgotten any special cases. 

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

gassmoeller commented on source/simulator/assembly.cc:
```diff
@@ -146,25 +146,40 @@ namespace aspect
+            assemblers->advection_system_on_interior_face[i].push_back(
+              std::make_unique<aspect::Assemblers::AdvectionSystemInteriorFace<dim>>());
+          }
+
```
below: `|| parameters.temperature_method != Parameters<dim>::AdvectionFieldMethod::fem_field`

gassmoeller reviewed (commented):
Looks correct, just one comment.

bangerth reviewed (commented):
I think this is the right approach: Select up front which assemblers to call, rather than calling them and letting them return right away.

OK if you can address @gassmoeller 's comment.

bangerth commented:
```
The following tests FAILED:
	225 - darcy_velocity (Failed)
	642 - particle_periodic_boundaries_dc (Failed)
```

jdannberg commented:
I figures out why the two tests were failing:

The Darcy flow one was an actual bug in the Darcy assembler. If that field method was selected, both the normal advcetion assembler and the Darcy assembler were called, so the result was wrong. I made a new test which is working now, but would give us the wrong solution before. The idea is that there is a blob, and the solid velocity is downwards, whereas the darcy flow is upwards, and the amplitudes are equal, so the blob should not move. 

Original blob location:
![start](https://github.com/geodynamics/aspect/assets/7517347/c46d5884-e3cb-49ff-8a74-65f2423f8a4d)

New solution: Blob stays in place. 
![end](https://github.com/geodynamics/aspect/assets/7517347/8ddd9d4a-b536-4939-8c18-d9fc29ba8da9)

Old (wrong) solution: Blob has moved down. 
![wrong](https://github.com/geodynamics/aspect/assets/7517347/5e831250-5c28-45a6-8730-14025de18d4c)

Since the change I made relies on being able to use different assemblers, I had to include it in this PR rather than making a separate one.

In addition, the assembler was not derived from the StabilizationAssemblerInterface, which I fixed now.


The other failing test was an Assert I changed so that it was triggering now. I fixed the Assert, but one of the problems was that both on my laptop and on the tester, we did not actually trigger it, but instead got this error message:

```
*** "Test particle_periodic_boundaries_dc failed": ***
[1/5] Generating output-particle_periodic_boundaries_dc/screen-output
FAILED: output-particle_periodic_boundaries_dc/screen-output 
cd /jenkins/workspace/aspect_PR-5242/build/tests && /jenkins/workspace/aspect_PR-5242/tests/cmake/timeout.sh 600s mpirun -np 2 /jenkins/workspace/aspect_PR-5242/build/aspect /jenkins/workspace/aspect_PR-5242/build/tests/particle_periodic_boundaries_dc.x.prm > /jenkins/workspace/aspect_PR-5242/build/tests/output-particle_periodic_boundaries_dc/screen-output.tmp && /jenkins/workspace/aspect_PR-5242/tests/cmake/default screen-output </jenkins/workspace/aspect_PR-5242/build/tests/output-particle_periodic_boundaries_dc/screen-output.tmp >/jenkins/workspace/aspect_PR-5242/build/tests/output-particle_periodic_boundaries_dc/screen-output

--------------------------------------------------------
An error occurred in line <384> of file </opt/tmp/unpack/deal.II-v9.4.1/include/deal.II/base/smartpointer.h> in function
    T* dealii::SmartPointer<T, P>::operator->() const [with T = const dealii::Triangulation<2, 2>; P = dealii::Particles::ParticleHandler<2, 2>]
The violated condition was: 
    pointed_to_object_is_alive
Additional information: 
    The object pointed to is not valid anymore.

Stacktrace:
-----------
#0  /opt/deal.II-v9.4.1/lib/libdeal_II.g.so.9.4.1: dealii::SmartPointer<dealii::Triangulation<2, 2> const, dealii::Particles::ParticleHandler<2, 2> >::operator->() const
#1  /opt/deal.II-v9.4.1/lib/libdeal_II.g.so.9.4.1: dealii::Particles::ParticleHandler<2, 2>::reset_particle_container(std::__cxx11::list<dealii::Particles::ParticleAccessor<2, 2>::ParticlesInCell, std::allocator<dealii::Particles::ParticleAccessor<2, 2>::ParticlesInCell> >&)
#2  /opt/deal.II-v9.4.1/lib/libdeal_II.g.so.9.4.1: dealii::Particles::ParticleHandler<2, 2>::clear_particles()
#3  /opt/deal.II-v9.4.1/lib/libdeal_II.g.so.9.4.1: dealii::Particles::ParticleHandler<2, 2>::~ParticleHandler()
#4  /jenkins/workspace/aspect_PR-5242/build/aspect: aspect::Particle::World<2>::~World()
#5  /jenkins/workspace/aspect_PR-5242/build/aspect: aspect::Particle::World<2>::~World()
#6  /jenkins/workspace/aspect_PR-5242/build/aspect: aspect::Simulator<2>::Simulator(ompi_communicator_t*, dealii::ParameterHandler&)
#7  /jenkins/workspace/aspect_PR-5242/build/aspect: void run_simulator<2>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool, bool)
#8  /jenkins/workspace/aspect_PR-5242/build/aspect: main
--------------------------------------------------------

Calling MPI_Abort now.
To break execution in a GDB session, execute 'break MPI_Abort' before running. You can also put the following into your ~/.gdbinit:
  set breakpoint pending on
  break MPI_Abort
  set breakpoint pending auto
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 1 in communicator MPI_COMM_WORLD
with errorcode 255.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
[2896cf7c666a:12146] PMIX ERROR: UNREACHABLE in file ../../../src/server/pmix_server.c at line 2193

--------------------------------------------------------
An error occurred in line <384> of file </opt/tmp/unpack/deal.II-v9.4.1/include/deal.II/base/smartpointer.h> in function
    T* dealii::SmartPointer<T, P>::operator->() const [with T = const dealii::Triangulation<2, 2>; P = dealii::Particles::ParticleHandler<2, 2>]
The violated condition was: 
    pointed_to_object_is_alive
Additional information: 
    The object pointed to is not valid anymore.

Stacktrace:
-----------
#0  /opt/deal.II-v9.4.1/lib/libdeal_II.g.so.9.4.1: dealii::SmartPointer<dealii::Triangulation<2, 2> const, dealii::Particles::ParticleHandler<2, 2> >::operator->() const
#1  /opt/deal.II-v9.4.1/lib/libdeal_II.g.so.9.4.1: dealii::Particles::ParticleHandler<2, 2>::reset_particle_container(std::__cxx11::list<dealii::Particles::ParticleAccessor<2, 2>::ParticlesInCell, std::allocator<dealii::Particles::ParticleAccessor<2, 2>::ParticlesInCell> >&)
#2  /opt/deal.II-v9.4.1/lib/libdeal_II.g.so.9.4.1: dealii::Particles::ParticleHandler<2, 2>::clear_particles()
#3  /opt/deal.II-v9.4.1/lib/libdeal_II.g.so.9.4.1: dealii::Particles::ParticleHandler<2, 2>::~ParticleHandler()
#4  /jenkins/workspace/aspect_PR-5242/build/aspect: aspect::Particle::World<2>::~World()
#5  /jenkins/workspace/aspect_PR-5242/build/aspect: aspect::Particle::World<2>::~World()
#6  /jenkins/workspace/aspect_PR-5242/build/aspect: aspect::Simulator<2>::Simulator(ompi_communicator_t*, dealii::ParameterHandler&)
#7  /jenkins/workspace/aspect_PR-5242/build/aspect: void run_simulator<2>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool, bool)
#8  /jenkins/workspace/aspect_PR-5242/build/aspect: main
--------------------------------------------------------

Calling MPI_Abort now.
To break execution in a GDB session, execute 'break MPI_Abort' before running. You can also put the following into your ~/.gdbinit:
  set breakpoint pending on
  break MPI_Abort
  set breakpoint pending auto
[2896cf7c666a:12146] 1 more process has sent help message help-mpi-api.txt / mpi-abort
[2896cf7c666a:12146] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
ninja: build stopped: subcommand failed.

CMake Error at /jenkins/workspace/aspect_PR-5242/tests/run_test.cmake:14 (MESSAGE):
  *** Test aborted.
```

I would just open an issue about that, since I have no idea how to fix it, but it should not be a problem for this PR.

bangerth commented:
About the bug: oops :-)

About the assertion: This happens because there's an exception somewhere that is not caught. The exception propagates upwards, and in the process of destroying objects, we come to a point where one object references another that no longer exists. If you have a test case that illustrates this, then I will fix it for you.

bangerth commented:
There are some tests to fix, though:
```
FAILED: output-darcy_velocity_vertical/screen-output 
cd /__w/aspect/aspect/build/tests && /__w/aspect/aspect/tests/cmake/timeout.sh 600s /__w/aspect/aspect/tests/cmake/run_test.sh 0 /__w/aspect/aspect/build/aspect /__w/aspect/aspect/build/tests/darcy_velocity_vertical.x.prm /__w/aspect/aspect/build/tests/output-darcy_velocity_vertical/screen-output.tmp && /__w/aspect/aspect/tests/cmake/default screen-output </__w/aspect/aspect/build/tests/output-darcy_velocity_vertical/screen-output.tmp >/__w/aspect/aspect/build/tests/output-darcy_velocity_vertical/screen-output
-----------------------------------------------------------------------------
-- This is ASPECT, the Advanced Solver for Problems in Earth's ConvecTion.
--     . version 2.6.0-pre
--     . using deal.II 9.4.1
--     .       with 32 bit indices and vectorization level 1 (128 bits)
--     . using Trilinos 13.2.0
--     . using p4est 2.3.2
--     . running in DEBUG mode
--     . running with 1 MPI process
-----------------------------------------------------------------------------

    Line <121> of file </__w/aspect/aspect/tests/darcy_velocity.prm>:
    The entry value 
        melting rate
    for the entry named
        Model name
    does not match the given pattern:
        [Selection Steinberger|ascii reference profile|averaging|compositing|composition reaction|depth dependent|diffusion dislocation|drucker prager|grain size|latent heat|latent heat melt|melt boukare|melt global|melt simple|modified tait|multicomponent|multicomponent compressible|nondimensional|perplex lookup|replace lithosphere viscosity|simple|simple compressible|simpler|visco plastic|viscoelastic|unspecified ]



----------------------------------------------------
Exception 'dealii::ExcMessage ("Invalid input parameter file.")' on rank 0 on processing: 

--------------------------------------------------------
An error occurred in line <[479](https://github.com/geodynamics/aspect/actions/runs/5515036438/jobs/10054946343?pr=5242#step:6:480)> of file </__w/aspect/aspect/source/main.cc> in function
    void parse_parameters(const string&, dealii::ParameterHandler&)
The violated condition was: 
    false
Additional information: 
    Invalid input parameter file.

Stacktrace:
-----------
#0  /__w/aspect/aspect/build/aspect: parse_parameters(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, dealii::ParameterHandler&)
#1  /__w/aspect/aspect/build/aspect: void run_simulator<2>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool, bool)
#2  /__w/aspect/aspect/build/aspect: main
--------------------------------------------------------

Aborting!
----------------------------------------------------
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
ninja: build stopped: subcommand failed.

CMake Error at /__w/aspect/aspect/tests/run_test.cmake:14 (MESSAGE):
  *** Test aborted.
```

and

```
FAILED: output-multiple_named_additional_outputs/screen-output 
cd /__w/aspect/aspect/build/tests && /__w/aspect/aspect/tests/cmake/timeout.sh 600s /__w/aspect/aspect/tests/cmake/run_test.sh 0 /__w/aspect/aspect/build/aspect /__w/aspect/aspect/build/tests/multiple_named_additional_outputs.x.prm /__w/aspect/aspect/build/tests/output-multiple_named_additional_outputs/screen-output.tmp && /__w/aspect/aspect/tests/cmake/default screen-output </__w/aspect/aspect/build/tests/output-multiple_named_additional_outputs/screen-output.tmp >/__w/aspect/aspect/build/tests/output-multiple_named_additional_outputs/screen-output
/__w/aspect/aspect/tests/cmake/run_test.sh: line 16: 50724 Aborted                 (core dumped) $BINARY $PRM > ${OUTPUT} 2>&1
-----------------------------------------------------------------------------
-- This is ASPECT, the Advanced Solver for Problems in Earth's ConvecTion.
--     . version 2.6.0-pre
--     . using deal.II 9.4.1
--     .       with 32 bit indices and vectorization level 1 (128 bits)
--     . using Trilinos 13.2.0
--     . using p4est 2.3.2
--     . running in DEBUG mode
--     . running with 1 MPI process
-----------------------------------------------------------------------------

Loading shared library <./libmultiple_named_additional_outputs.debug.so>

-----------------------------------------------------------------------------
-- For information on how to cite ASPECT, see:
--   https://aspect.geodynamics.org/citing.html?ver=2.6.0-pre&sha=&src=code
-----------------------------------------------------------------------------
Number of active cells: 400 (on 2 levels)
Number of degrees of freedom: 10,628 (4,010+603+2,005+2,005+2,005)

*** Timestep 0:  t=0 seconds, dt=0 seconds
   Solving temperature system... 0 iterations.
   Skipping normal_field composition solve because RHS is zero.
   Copying properties into prescribed compositional field diffused_field... done.

--------------------------------------------------------
An error occurred in line <146> of file </__w/aspect/aspect/source/simulator/entropy_viscosity.cc> in function
    double aspect::Simulator<dim>::compute_viscosity(aspect::internal::Assembly::Scratch::AdvectionSystem<dim>&, double, double, double, double, double, const aspect::Simulator<dim>::AdvectionField&) const [with int dim = 2]
The violated condition was: 
    stabilization_assembler != nullptr
Additional information: 
    Entropy viscosity can only be computed for assemblers that are derived
    from the advection stabilization assembler interface.
```
and perhaps others -- I didn't look further down the log files.

tjhei commented:
```
/__w/aspect/aspect/source/simulator/assemblers/advection.cc:399:5: error: 'std::vector<double> aspect::Assemblers::DiffusionSystem<dim>::advection_prefactors(aspect::internal::Assembly::Scratch::ScratchBase<dim>&) const [with int dim = 3]' can be marked override [-Werror=suggest-override]
```
